lib/dialects/abstract/query-generator/query-proto.js

```diff
@@ -0,0 +1,157 @@
+'use strict';
+
+const { Composition } = require('./composition');
+
+/*
+ * Named collections of Composition elements.
+ *
+ * Useful for organizing and manipulating the different parts of a query.
+ * Allow disordered query building and easy substitution of query parts. Flexible,
+ * non-nested and non-exahustive organization.
+ */
+
+class QueryProto {
+  constructor(proto) {
+    if (proto) {
+      this.set(proto);
+    } else {
+      // Should clone Composition, NOT copy shallow references
+      for (const key of this.constructor.partNames) {
+        this[key] = new Composition();
+      }
+    }
+  }
+  static get partNames() {
+    return [];
+  }
+  set(proto) {
+    // Should clone proto into this object, NOT copy shallow references
+
+    for (const key of this.constructor.partNames) {
+      this[key] = new Composition();
+
+      if (proto[key] && proto[key].length) {
+        this[key] = proto[key].clone();
+      } else {
+        this[key] = new Composition();
+      }
+    }
+
+    return this;
+  }
+  clone() {
+    // Should clone Composition, NOT copy shallow references
+    const clone = new this.constructor();
+
+    for (const key of this.constructor.partNames) {
+      clone[key] = this[key].clone();
+    }
+
+    return clone;
+  }
+  toComposition() {
+    throw new Error(`Class ${this.constructor.name} has not implemented toComposition method`);
+  }
+}
+
+class SelectProto extends QueryProto {
+  static get partNames() {
+    return Object.freeze(['attributes', 'from', 'join', 'where', 'group',
+      'having', 'order', 'page', 'lock']);
+  }
+  toComposition() {
+    const composition = new Composition();
+    composition.add('SELECT');
+
+    if (this.attributes.length) {
+      composition.add(' ', this.attributes);
+    } else {
+      composition.add(' *');
+    }
+
+    if (! (this.from && this.from.length)) {
+      throw new Error('Missing FROM clause in select proto query');
+    }
+
+    composition.add(' FROM ', this.from);
+
+    // Each join should come with its own join linker
+    if (this.join.length) composition.add(' ', this.join);
+    if (this.where.length) composition.add(' WHERE ', this.where);
+    if (this.group.length) composition.add(' GROUP BY ', this.group);
+    if (this.having.length) composition.add(' HAVING ', this.having);
+    if (this.order.length) composition.add(' ORDER BY ', this.order);
+    // Each dialect has its own pagination and lock syntax
+    if (this.page.length) composition.add(' ', this.page);
+    if (this.lock.length) composition.add(' ', this.lock);
+
+    return composition;
+  }
+}
+
+class InsertProto extends QueryProto {
+  static get partNames() {
+    return Object.freeze(['preQuery', 'flags', 'table', 'ignoreDups', 'attributes',
+      'output', 'values', 'onConflict', 'return', 'postQuery']);
+  }
+  toComposition() {
+    if (!this.table || !this.table.length) {
+      throw new Error('Missing table in INSERT query');
```

Comment:
Shouldn't this throw the `CompositionError` you created?

Reply:
Yes, now it uses it